src/vmm/src/signal_handler.rs

```diff
@@ -49,62 +83,88 @@ extern "C" fn sigsys_handler(num: c_int, info: *mut siginfo_t, _unused: *mut c_v
         unsafe { _exit(i32::from(super::FC_EXIT_CODE_UNEXPECTED_ERROR)) };
     }
 
     // Other signals which might do async unsafe things incompatible with the rest of this
     // function are blocked due to the sa_mask used when registering the signal handler.
-    match si_signo {
-        SIGBUS => METRICS.signals.sigbus.inc(),
-        SIGSEGV => METRICS.signals.sigsegv.inc(),
-        _ => (),
-    }
-
+    let syscall = unsafe { *(info as *const i32).offset(SI_OFF_SYSCALL) as usize };
     error!(
-        "Shutting down VM after intercepting signal {}, code {}.",
-        si_signo, si_code
+        "Shutting down VM after intercepting a bad syscall ({}).",
+        syscall
     );
-    // Write the metrics before exiting.
-    if let Err(e) = METRICS.write() {
-        error!("Failed to write metrics while stopping: {}", e);
-    }
-
-    // Safe because we're terminating the process anyway. We don't actually do anything when
-    // running unit tests.
-    #[cfg(not(test))]
-    unsafe {
-        _exit(i32::from(match si_signo {
-            SIGBUS => super::FC_EXIT_CODE_SIGBUS,
-            SIGSEGV => super::FC_EXIT_CODE_SIGSEGV,
-            _ => super::FC_EXIT_CODE_UNEXPECTED_ERROR,
-        }))
-    };
 }
 
+fn empty_fn(_si_code: c_int, _info: *mut siginfo_t) {}
+
+generate_handler!(
+    sigxfsz_handler,
+    SIGXFSZ,
+    FC_EXIT_CODE_SIGXFSZ,
+    METRICS.signals.sigxfsz,
+    empty_fn
+);
+
+generate_handler!(
+    sigxcpu_handler,
+    SIGXCPU,
+    FC_EXIT_CODE_SIGXCPU,
+    METRICS.signals.sigxcpu,
+    empty_fn
+);
+
+generate_handler!(
+    sigbus_handler,
+    SIGBUS,
+    FC_EXIT_CODE_SIGBUS,
+    METRICS.signals.sigbus,
+    empty_fn
+);
+
+generate_handler!(
+    sigsegv_handler,
+    SIGSEGV,
+    FC_EXIT_CODE_SIGSEGV,
+    METRICS.signals.sigsegv,
+    empty_fn
+);
+
+generate_handler!(
+    sigpipe_handler,
+    SIGPIPE,
+    FC_EXIT_CODE_SIGPIPE,
+    METRICS.signals.sigpipe,
+    empty_fn
+);
+
+generate_handler!(
+    sigsys_handler,
+    SIGSYS,
+    FC_EXIT_CODE_BAD_SYSCALL,
+    METRICS.seccomp.num_faults,
+    log_sigsys_err
+);
+
 /// Registers all the required signal handlers.
 ///
-/// Custom handlers are installed for: `SIGBUS`, `SIGSEGV`, `SIGSYS`.
+/// Custom handlers are installed for: `SIGBUS`, `SIGSEGV`, `SIGSYS`
+/// `SIGXFSZ` `SIGXCPU` and `SIGPIPE` .
 pub fn register_signal_handlers() -> utils::errno::Result<()> {
     // Call to unsafe register_signal_handler which is considered unsafe because it will
     // register a signal handler which will be called in the current thread and will interrupt
     // whatever work is done on the current thread, so we have to keep in mind that the registered
     // signal handler must only do async-signal-safe operations.
     register_signal_handler(SIGSYS, sigsys_handler)?;
-    register_signal_handler(SIGBUS, sigbus_sigsegv_handler)?;
-    register_signal_handler(SIGSEGV, sigbus_sigsegv_handler)?;
+    register_signal_handler(SIGBUS, sigbus_handler)?;
```

Comment:
Apart from these signals, I would also add to this list the SIGHUP (for when the terminal where firecracker started vanishes) and SIGILL (for when users would try to run firecracker on the wrong architecture, or in case of corrupted binary)  signals.